File: 3-leetcode/01-easy/js/530-minimum-absolute-difference-in-bst.ts
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

const getMinimumDifference = (root: TreeNode | null): number => {
  let minimum = Number.MAX_SAFE_INTEGER;
  const nums = [];

  const stack = [root];

  while (stack.length) {
    const cur = stack.pop();
    nums.push(cur.val);
    if (cur.left) {
      stack.push(cur.left);
    }
    if (cur.right) {
      stack.push(cur.right);
    }
  }

  nums.sort((a, b) => a - b);

  const length = nums.length;

  for (let i = 0; i < length - 1; i++) {
    minimum = Math.min(Math.abs(nums[i] - nums[i + 1]), minimum);
  }

  return minimum;
};

File: 3-leetcode/01-easy/cpp/530-minimum-absolute-difference-in-bst.cpp
#include <bits/stdc++.h>

using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    int diff = INT_MAX;
    TreeNode *prev = NULL;

    void dfs(TreeNode *node)
    {
        if (node->left)
        {
            dfs(node->left);
        }

        if (prev)
        {
            diff = min(diff, abs(prev->val - node->val));
        }

        prev = node;

        if (node->right)
        {
            dfs(node->right);
        }
    }

    int getMinimumDifference(TreeNode *root)
    {
        dfs(root);
        return diff;
    }
};